src/grpc/stream_infer_handler.h

```diff
@@ -114,6 +115,20 @@ class ModelStreamInferHandler
       void* userp);
   bool Finish(State* state);
 
+  // Step helpers
+  bool RequestStartStep(State* state, bool rpc_ok);
+  bool RequestReadStep(State* state, bool rpc_ok);
+  bool RequestCompleteStep(State* state);
+  bool RequestWrittenStep(State* state, bool rpc_ok);
```

Comment:
RequestWrittenStep and RequestWrittenStepNonDecoupled looks confusing.
I would recommend pulling the state->is_decoupled_ out into the Process function. It might be easier to look at which function is being called when in the Process function.